`LinkNeighborLoader`: Support `data` with no edges.

wsad1 opened the pull request:
Sometimes we have `HeteroData` objects where there's just one `node_type` and we'd like to predict links between nodes of the same type. With this PR `LinkNeighborLoader` supports this.

What changes does this PR make
1. defines `LinkSampler` which is used instead of `LinkNeighborSampler` when `data` doesn't have edges. The `__call__` simply returns the indices to select in the `filter_fn`.
2. defines `LinkSamplerMixin` which contains common function between `LinkSampler` and `LinkNeighborSampler`.

Padarn commented on torch_geometric/loader/link_neighbor_loader.py:
```diff
@@ -258,7 +305,12 @@ def __init__(
+        # Corner Case: When data doesn't have any edges
+        # return a subset of nodes.
+        if not has_edge_index(data):
+            self.neighbor_sampler = LinkSampler(data, neg_sampling_ratio,
```
```suggestion
            self.neighbor_sampler = LinkSampler(data, self.neg_sampling_ratio,
```
or modify it below, just for consistency

Padarn commented on torch_geometric/loader/link_neighbor_loader.py:
```diff
@@ -52,43 +54,88 @@ def _create_label(self, edge_label_index, edge_label):
+        super().__init__()
+        self.data_cls = data.__class__ if isinstance(
+            data, (Data, HeteroData)) else 'custom'
+        self.neg_sampling_ratio = neg_sampling_ratio
```
is this used in the `LinkSampler`?

wsad1 replied:
Do you mean the `neg_samling_ratio` or `data_cls`?
`data_cls` is used in the `__call__`.

Padarn commented on torch_geometric/loader/utils.py:
```diff
@@ -224,3 +224,21 @@ def filter_feature_store(
     return data
+
+
+def has_edge_index(data: Union[Data, HeteroData]) -> bool:
```
Could this be defined on `Data`?

wsad1 replied:
Yeah we could add it to `HeteroData`, `Data` and probably `GraphStore`. That's much cleaner. But I doubt it would be used outside of `LinkNeighborLoader`.

Padarn commented on torch_geometric/loader/link_neighbor_loader.py:
```diff
@@ -258,7 +305,12 @@ def __init__(
         edge_type, edge_label_index = get_edge_label_index(
```
I think this might break if your graph has no edges?

wsad1 replied:
I do have tests for graphs without edges, it works fine for them. Do you have some specific case in mind

Padarn replied:
sorry I meant if `edge_label_index = None` (default)

wsad1 replied:
Good point, I think we should force the user the pass an `edge_label_index` if the graph doesn't have edges. Maybe with an assert statement.

Padarn commented:
@wsad1 I like the refactoring a lot, thanks!

I don't really understand the use case though sorry (I'm just not familiar enough) do you have a dataset/task you can refer me to so I can understand better?

wsad1 commented:
> @wsad1 I like the refactoring a lot, thanks!
> 
> I don't really understand the use case though sorry (I'm just not familiar enough) do you have a dataset/task you can refer me to so I can understand better?

@Padarn the test I added ` test_hetero_no_edge_graph_link_neighbor_loader` is this PR contains a use case. In the test,  we have a graph with `paper` nodes and no edges and we'd like to predict links between `paper` nodes. Previously the `LinkNeighborLoader` wouldn't work as there aren't any edges (neighbors). This PR is kind of a hack to support this.

Padarn commented:
thanks... after looking at the the example its more clear!

wsad1 commented:
Sorry I have to close this PR as its very hard to resolve some conflicts. Closing in favour of #5061.
@Padarn, I have addressed some of your comments in the new PR.
